add flake8 pre-commit

Zethson opened the pull request:
I intervened a couple of times manually and switched off a couple of checks, but it should be a strong improvement.

Signed-off-by: Zethson <lukas.heumos@posteo.net>

flying-sheep committed:
Switch to flit

This reverts commit d6457902

flying-sheep committed:
Update install instructions

flying-sheep committed:
add setup.py while leaving it ignored

flying-sheep committed:
Circumvent new pip check (see pypa/pip#9628)

Zethson committed:
add flake8 pre-commit

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add flake8 pre-commit

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix pre-commit

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson commented:
`E   ImportError: cannot import name 'settings' from partially initialized module 'scanpy' (most likely due to a circular import) (/home/vsts/work/1/s/scanpy/__init__.py)
`

Meh, it's hell to track this down now. I assume that autopep8 removed an unused variable.

Zethson committed:
add E402 to flake8 ignore

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add E402 to flake8 ignore

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
revert neighbors

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
revert neighbors

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
Merge branch 'master' into feature/flake8

Zethson committed:
fix flake8

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson commented:
```
$ python -m scanpy.tests.blackdiff 10

/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/setuptools_scm/git.py:68: UserWarning: "/home/travis/build/theislab/scanpy" is shallow and may cause errors

  warnings.warn('"{}" is shallow and may cause errors'.format(wd.path))

/home/travis/virtualenv/python3.7.1/bin/python: No module named scanpy.tests.blackdiff
```

Not sure why this happened, but we well we're using black with pre-commit now anyways so w/e. Does this need fixing?

Koncopd committed:
rm call of black test (#1690)

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,4 @@
+[flake8]
+exclude = docs, scanpy/tests
+max-line-length = 120
+ignore = F401, W503, E501, E203, E231, W504, E402, E126, E712, E741
```
For checks that are switched off, could there be some documentation about why? If it's taken from a template it'd be fine to just link to where that came from.

ivirshup commented on pyproject.toml:
```diff
@@ -26,8 +26,8 @@ source = ['scanpy']
 [tool.black]
-line-length = 88
-target-version = ['py36']
+line-length = 120
```
This shouldn't change, it's definitely out of scope for the pr, and I think 120 is too long.

It also makes reviewing the diffs in this PR difficult. Could you change this and I'll come back for a review?

Zethson replied:
Then I guess this will a follow up discussion :)

88 makes for some really ugly code imo and 120 is not a lot. But will see what everyone says :)

ivirshup replied:
Haha, last time we went through this I was the one pushing hard for increasing it from 80 😆

ivirshup reviewed (changes requested):
Main thing: Can you get rid of the line length change and re-request review?

-----------------------

> scanpy.tests.blackdiff

This file got deleted, so I'm actually kinda surprised you're only getting a warning. Whatever this check is can be removed.

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,4 @@
+[flake8]
+exclude = docs, scanpy/tests
```
Why exclude docs and tests?

I expect some pathological things in tests (unused imports for pytest fixtures), can we disable specific checks by path?

Zethson replied:
Oh, I did this before to get at least some sanity. There was so much stuff to fix that I excluded them and then forgot.

https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores

ivirshup commented on .pre-commit-config.yaml:
```diff
@@ -3,3 +3,8 @@ repos:
+    rev: 3.8.4
+    hooks:
+    -   id: flake8
+        exclude: scanpy/tests/
```
Can we just have exclude be defined in one place?

Zethson replied:
No, unfortunately not. pre-commit ignores excludes in the .flake8 file. 

https://github.com/pre-commit/pre-commit-hooks/issues/112

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,4 @@
+[flake8]
```
A comment here about when this can be moved to the `pyproject.toml` would be useful. Could be [a link to this](https://gitlab.com/pycqa/flake8/-/issues/428#note_251982786)

ivirshup committed:
Fix print_versions for python<3.8 (#1691)

ivirshup committed:
add codecov so we can have a badge to point to (#1693)

ivirshup committed:
Attempt server-side search (#1672)

Zethson committed:
address review

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
address review

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix comment character in .flake8

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix test

Signed-off-by: Zethson <lukas.heumos@posteo.net>

ivirshup commented on scanpy/plotting/_utils.py:
```diff
@@ -619,7 +619,7 @@ def setup_axes(
     draw_region_width_frac = draw_region_width / figure_width
     left_offset_frac = left_offset / figure_width
-    right_offset_frac = 1 - (len(panels) - 1) * left_offset_frac
+    right_offset_frac = 1 - (len(panels) - 1) * left_offset_frac  # noqa: F841
```
?

ivirshup replied:
@fidelram, should this line be removed?

Zethson replied:
Yeah I could remove it of course, but I wanted this PR to be rather defensive about manual interventions of mine.

ivirshup commented on scanpy/plotting/_utils.py:
```diff
@@ -939,8 +937,8 @@ def make_pos(pos, node=root, currentLevel=0, parent=None, vert_loc=0):
-        levels = {l: {TOTAL: levels[l], CURRENT: 0} for l in levels}
-    vert_gap = height / (max([l for l in levels]) + 1)
+        levels = {level: {TOTAL: levels[level], CURRENT: 0} for level in levels}
+    vert_gap = height / (max([level for level in levels]) + 1)
```
I might want to disable this rule (which one is this?). Single letter variable names in comprehension expressions are fine.

Zethson replied:
It's not about the single letter, but the single letter reusage

ivirshup replied:
But variables defined inside generators aren't defined in outer scopes:

```python
In [1]: [[i for i in range(j)] for j in range(5)]
Out[1]: [[], [0], [0, 1], [0, 1, 2], [0, 1, 2, 3]]

In [2]: i
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-397d543883c5> in <module>
----> 1 i

NameError: name 'i' is not defined

In [3]: j
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-3eedd8854d1e> in <module>
----> 1 j

NameError: name 'j' is not defined
```

So, what issue is being avoided here?

Zethson replied:
E741. Even Pycharm flags it already.

ivirshup replied:
Ehh. If the issue is letters that look like numbers (a) in what programming font is this the case (b) they should also be different colors. I'd be for turning this rule off.

Zethson replied:
I would keep it since the IDE  + flake8 flag it.

Zethson replied:
It causes 0 harm to adhere to better naming practices.

ivirshup commented on scanpy/preprocessing/_deprecated/highly_variable_genes.py:
```diff
@@ -170,8 +170,10 @@ def filter_genes_dispersion(
+            df['dispersion'].values
+            - disp_mean_bin[
+                df['mean_bin'].values
+            ].values  # use values here as index differs
```
I think this comment ended up pointing at the wrong place.

Zethson replied:
True -> 2x black with different line numbers does this. black is not perfect.

ivirshup commented on scanpy/neighbors/__init__.py:
```diff
@@ -126,7 +124,6 @@ def neighbors(
     **distances** : sparse matrix of dtype `float32`.
         Instead of decaying weights, this stores distances for each pair of
         neighbors.
-    
```
I feel like there should be a blank line here. Was this a flake8 error?

ivirshup replied:
Did this get resolved? I'm not seeing an outdated.

Zethson replied:
Yeah, I added the line back

ivirshup commented on scanpy/plotting/_anndata.py:
```diff
@@ -1960,7 +1960,7 @@ def _plot_gene_groups_brackets(
                     rotation=rotation,
                 )
-            except:
+            except Exception:
```
Is this better? If anything, I feel like this should get an issue opened. Maybe a new tag for "scary code"?

Zethson replied:
No it's not better. It doesn't change anything semantically, but at least it makes flake8 shut up. This PR cannot fix all violations. Scanpy had like 200 flake8 violations. 

So +1 for collecting these in a new issue and fixing them over time, but this PR will not do that.

ivirshup replied:
Note to self: this conversation is being continued in the main discussion thread.

ivirshup commented on scanpy/plotting/_baseplot_class.py:
```diff
@@ -121,13 +121,13 @@ def __init__(
             if set(self.obs_tidy.index.categories) != set(categories_order):
                 logg.error(
-                    "Please check that the categories given by "
+                    "Please check that the categories given by "  # noqa: F821
```
I don't see how this line could give this code (undefined variable?)

Zethson replied:
I don't know either, but this is the result

```
./scanpy/plotting/_baseplot_class.py:124:26: F821 undefined name 'obs_tidy'
./scanpy/plotting/_baseplot_class.py:124:27: F821 undefined name 'obs_tidy'
```

ivirshup commented on scanpy/plotting/_baseplot_class.py:
```diff
@@ -121,13 +121,13 @@ def __init__(
                     "want to be reordered.\n\n"
                     "Mismatch: "
-                    f"{set(obs_tidy.index.categories).difference(categories_order)}\n\n"
+                    f"{set(obs_tidy.index.categories).difference(categories_order)}\n\n"  # noqa: F821
```
Could you also explain these two?

Zethson replied:
See above.

ivirshup replied:
I expect these need `self.obs_tidy` since no variable is called `obs_tidy`

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -56,7 +56,7 @@ def pca_overview(adata: AnnData, **params):
     if 'show' in params:
         del params['show']
-    scatterplots.pca(adata, **params, show=False)
+    scatterplots.pca(adata, **params, show=False)  # noqa: F821
```
Also, how is this F821?

If it wants `scatterplots.pca(adata, show=False, **params)`, that seems fine.

Zethson replied:
`./scanpy/plotting/_tools/__init__.py:59:5: F821 undefined name 'scatterplots'`

Removing the `scatterplots` makes it happy. We're importing pca at the top anyways.

ivirshup replied:
My guess  is the intent here was to make it absolutely clear this isn't the `sc.tl.pca`. If we don't import `scatterplots.pca` here, does the warning still come up?

Zethson replied:
Yes.

ivirshup commented on scanpy/plotting/_tools/scatterplots.py:
```diff
@@ -252,7 +252,7 @@ def embedding(
         )
 
-        ### Order points
+        # Order points
```
If you're getting this, could you also get rid of the `###` on line 223?

But also, what rule is this?

Zethson replied:
./scanpy/plotting/_tools/scatterplots.py:254:9: E266 too many leading '#' for block comment

ivirshup replied:
I'd be for turning off E266 then. It's a legitimate way to style comments into sections. Even if it wasn't, this check is not catching bugs.

ivirshup commented on scanpy/plotting/_tools/scatterplots.py:
```diff
@@ -977,7 +976,7 @@ def _get_data_points(
             for comp in components_list:
                 data_points.append(adata.obsm[basis_key][:, comp])
-        except:
+        except Exception:
```
Again, I'm not sure `except Exception` is actually better than a bare `except:`.

Zethson replied:
It's not. But it makes flake8 happy. You can collect all of them and someone has to fix them later. I did not want to change the semantics.

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -903,9 +899,9 @@ def _detect_branching_single_haghverdi16(self, Dseg, tips):
         # permutations of tip cells
         ps = [
             [0, 1, 2],  # start by computing distances from the first tip
-            [1, 2, 0],  #             -"-                       second tip
```
I think the formatting was better before. What code is this?

Zethson replied:
./scanpy/tools/_dpt.py:902:25: E262 inline comment should start with '# '
./scanpy/tools/_dpt.py:904:12: E262 inline comment should start with '# '

ivirshup replied:
Lets turn off E262 and revert this then.

ivirshup commented on scanpy/tests/test_pca.py:
```diff
@@ -40,7 +38,7 @@
 
 
-def test_pca_transform(array_type):
+def test_pca_transform(array_type):  # noqa: F811
```
What's the error? I don't see how this fits: `redefinition of unused name from line N`

Zethson replied:
./scanpy/tests/test_pca.py:41:24: F811 redefinition of unused 'array_type' from line 6

ivirshup replied:
Can we turn of rules by module? I expect this will come up a lot and I don't want to add this to every test that uses a fixture.

Zethson replied:
per-file-ignores =
    module: rule

Feel free to push to this branch which rules you want to turn of for whatever module.

ivirshup commented on scanpy/preprocessing/_highly_variable_genes.py:
```diff
@@ -237,8 +236,10 @@ def _highly_variable_genes_single_batch(
+            df['dispersions'].values
+            - disp_mean_bin[
+                df['mean_bin'].values
+            ].values  # use values here as index differs
```
Moving comment

ivirshup commented on scanpy/preprocessing/_simple.py:
```diff
@@ -757,7 +757,9 @@ def scale(
-    return scale_array(data, zero_center=zero_center, max_value=max_value, copy=copy)
+    return scale_array(
+        data, zero_center=zero_center, max_value=max_value, copy=copy  # noqa: F821
+    )
```
Wait, this can't be right. I guess it's just never being called?

Zethson replied:
It doesn't know `data`. Even Pycharm flags this. I don't know how such code even ended up in Scanpy.

ivirshup replied:
For this review process, if there is code that looks bad I would prefer that you flagged it for removal rather than adding it to an ignore list.

I think this is where a lot of the value of turning `flake8` on will come from.

Zethson replied:
Sure. But then you will have to fix it **now** or flake8 will keep failing :)

Whatever you prefer.

giovp replied:
but is this the `sc.pp.scale` ?  I wonder if it's tested?

ivirshup commented on scanpy/tools/_rank_genes_groups.py:
```diff
@@ -35,22 +35,32 @@ def _ranks(X, mask=None, mask_rest=None):
+
     else:
         n_cells = X.shape[0]
-        get_chunk = lambda X, left, right: adapt(X[:, left:right])
```
We should turn this one off.

Zethson replied:
The lambdas are not the issue. This was likely an unused variable.

ivirshup replied:
I'm not sure I understand the change then. How does replacing a `lambda` with a `def` solve this?

Zethson replied:
I was wrong earlier.

```
./scanpy/tools/_rank_genes_groups.py:38:9: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:39:9: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:42:9: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:48:9: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:53:9: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:171:13: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:173:13: E731 do not assign a lambda expression, use a def
./scanpy/tools/_rank_genes_groups.py:743:13: E731 do not assign a lambda expression, use a def
```

ivirshup replied:
I'm for turning off `E731`.

Zethson replied:
Context:
```

    The first form means that the name of the resulting function object is specifically 'f' instead of the generic '<lambda>'. This is more useful for tracebacks and string representations in general. The use of the assignment statement eliminates the sole benefit a lambda expression can offer over an explicit def statement (i.e. that it can be embedded inside a larger expression)

Assigning lambdas to names basically just duplicates the functionality of def - and in general, it's best to do something a single way to avoid confusion and increase clarity.

The legitimate use case for lambda is where you want to use a function without assigning it, e.g:
```

I like lambdas as well so you can make the final call.

ivirshup commented on scanpy/plotting/_utils.py:
```diff
@@ -742,9 +742,7 @@ def scatter_base(
-            cb = pl.colorbar(
-                sct, format=ticker.FuncFormatter(ticks_formatter), cax=ax_cb
-            )
+            pl.colorbar(sct, format=ticker.FuncFormatter(ticks_formatter), cax=ax_cb)
```
Why did this line change? I assume black would get this.

Zethson replied:
unused variable

ivirshup replied:
Ah, didn't see that

Why is this sometimes changed to `_ = statement` in other places? Was this an automatic tool, or manual?

Zethson replied:
Manual.

_ is the consensus for variables that get thrown away.

__ is not the standard -> I had to rename those as well

ivirshup commented on scanpy/tests/conftest.py:
```diff
@@ -10,6 +10,7 @@
 
 import scanpy
 
+
```
Is this a different rule than black?

Zethson replied:
I used autopep8 to get rid of many violations. Could be that it introduced this line

ivirshup commented on .pre-commit-config.yaml:
```diff
@@ -3,3 +3,8 @@ repos:
+    rev: 3.8.4
+    hooks:
+    -   id: flake8
+        exclude: scanpy/tests/
```
Aren't you including the tests now?

ivirshup commented on pyproject.toml:
```diff
@@ -27,7 +27,7 @@ omit = ['*/tests/*']
 [tool.black]
 line-length = 88
-target-version = ['py36']
+target-version = ['py38']
```
This should be the minimum python version we accept, no? Also, still out of scope. This is already touching so many things I'd prefer that there were no extraneous changes for ease of review.

ivirshup reviewed (changes requested):
Left a fair amount of comments and definitely didn't get everything.


## Commit process

I think this review process will be easier if you can limit changes from one flake error type to a single commit. 

This will make it easier to:

* Evaluate rules on their own. Sometimes rules aren't totally clear. I'm quite confused by some of the times you've used `# noqa: F821`
* Lets me know what rule the line was violating
* Revert changes if we decide we want to deactivate a particular rule. There are a few instances of this already, and I'd like to minimize the pain of this process.

This may be kinda hard to do now, but does this sound like a reasonable way to go forward?

## Rules I don't like

* I don't like replacing `x == False` with `not x` in all cases. Sometimes a variable could be a container, and an error should be thrown. I think cases have to be evaluated for this.
* Whats with changing from single letter variables inside expressions? Seems fine to me.
* `lambda's also are generally fine.
* Whats up with removing leading `#`s from comments?

## Process for bugs

So, some of the things you've adding a `# noqa` to look like bugs. I think we need to have a plan in place for doing something about these. Do you have any suggestions?

Zethson commented:
About the commit process: That's far far too much work to do it like you suggested. I don't have the time for this. 

About the rules: 

1. "I don't like replacing `x == False` with `not x` in all cases. Sometimes a variable could be a container, and an error should be thrown. I think cases have to be evaluated for this." 

This should be covered by tests. In any case it is not good style and a violation.

2. "Whats with changing from single letter variables inside expressions? Seems fine to me."

They are redefinitions of earlier variables and trip up flake8. We can call them whatever we want as long it s not `l` again. 

3. "`lambda's also are generally fine."

See comment at the section.

4. "Whats up with removing leading `#`s from comments?" Not my choice either. What we have now is pep8 and flake8 compliant. If you're not happy with this we can ignore the rule.

5. "So, some of the things you've adding a `# noqa` to look like bugs. I think we need to have a plan in place for doing something about these. Do you have any suggestions?"

The noqa ignore a rule for a specific line. I did not want to "fix" these things myself since Python is a dynamic language and you never know what happens :) Ideally we eventually get rid of all noqas, but not in this PR and not by me. I don't know the internals well enough to know whether this could have any side effects.

Zethson committed:
black

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
black

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
review round 2

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson commented:
@ivirshup tried to address all comments and made changes when necessary.

Zethson committed:
review round 3

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
readded double comments

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson commented:
@ivirshup I would keep the noqas. They are very easily searchable across the whole project and can be fixed later.

To me the PR is fine now. Any final wishes?

ivirshup commented:
> About the commit process: That's far far too much work to do it like you suggested. I don't have the time for this.

As a general point about this PR: to me, the fair amount of the work of turning on flake8 deciding on the rules. Perhaps we should start with a subset of files then? I realize you did not come to the meeting where we talked about this, so perhaps there is a difference of expectations here, but we agreed to be conservative about the rules we turned on in `pre-commit`.

Going through everything to make sure changes are correctly reverted is also takes a lot of time for me as the reviewer. I'd also like to limit that.

----------------

You said you used some automated tools to get faster compliance. What were these? In general, I would prefer to have a formatter that automatically ran than a tool that told me I formatted something wrong.

-----------------

> `@ivirshup` I would keep the noqas. They are very easily searchable across the whole project and can be fixed later.

I'm pretty strongly against this. `noqa`s just look like the formatter/ linter was wrong, and I'm not accepting that having no plan to address bugs. I think this should be a discussion with a broader set of the team.

> "Whats up with removing leading #s from comments?" Not my choice either. What we have now is pep8 and flake8 compliant. If you're not happy with this we can ignore the rule.

Yes, lets ignore this.

>> "I don't like replacing x == False with not x in all cases. Sometimes a variable could be a container, and an error should be thrown. I think cases have to be evaluated for this."
>
> This should be covered by tests. In any case it is not good style and a violation.

I will try and take a closer look at these changes. I'm particularly concerned that there will be cases where possible values are `None`, `True`, and `False`.

Zethson commented:
> You said you used some automated tools to get faster compliance. What were these? In general, I would prefer to have a formatter that automatically ran than a tool that told me I formatted something wrong.

https://pypi.org/project/autopep8/

Someone suggested to use it for pre-commit anyways. However, there is no full flake8 autofixer.


> As a general point about this PR: to me, the fair amount of the work of turning on flake8 deciding on the rules. Perhaps we should start with a subset of files then? 

Possible, but this will be an iterative process that will take a long time. If you are willing to do that I can close this PR and enable flake8 for example on the first folder of Scanpy. I cannot however fix all of them manually in like 20 PRs or something.

> I'm pretty strongly against this. noqas just look like the formatter/ linter was wrong, and I'm not accepting that having no plan to address bugs. I think this should be a discussion with a broader set of the team.

Depends on the view that you have. I see them more as TODOs for later since the "bad code" is in master at the moment anyways. But if you want to discuss that - sure.

Zethson committed:
Ignoring E262 & reverted comment

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
using self for obs_tidy

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
using self for obs_tidy

Signed-off-by: Zethson <lukas.heumos@posteo.net>

giovp commented on .flake8:
```diff
@@ -0,0 +1,5 @@
+# Can't yet be moved to the pyproject.toml due to https://gitlab.com/pycqa/flake8/-/issues/428#note_251982786
+[flake8]
+max-line-length = 88
```
I am also more happy with max line legnth=120. We have that in squidpy as well

Zethson replied:
I opened https://github.com/theislab/scanpy/issues/1694 to discuss this. Insert Bernie meme

giovp commented on scanpy/plotting/_matrixplot.py:
```diff
@@ -219,7 +219,7 @@ def _mainplot(self, ax):
             linewidth=self.edge_lw,
             norm=normalize,
         )
-        __ = ax.pcolor(_color_df, **kwds)
```
never seen `__` before 😅

giovp commented on scanpy/plotting/_utils.py:
```diff
@@ -126,7 +126,7 @@ def timeseries_subplot(
         levels, _ = np.unique(color, return_inverse=True)
         colors = np.array(palette[: len(levels)].by_key()['color'])
-        subsets = [(x_range[color == l], X[color == l, :]) for l in levels]
+        subsets = [(x_range[color == level], X[color == level, :]) for level in levels]
```
much better

giovp commented on scanpy/tools/_sim.py:
```diff
@@ -410,9 +409,8 @@ def __init__(
-        read = False
         if model not in self.availModels.keys():
-            message = 'model not among predefined models \n'
+            message = 'model not among predefined models \n'  # noqa: F841
```
this is weird

Zethson replied:
Yeah, it can absolutely be removed I guess :)
I didn't want to do it since I did not know whether you have some magic going on somewhere that looks for these "message" strings to log them or something.
But I guess we should remove or replace this with a log statement/exception here

giovp commented:
@Zethson thanks 💯 ! looks really good! Tbh I was expecting much worse, the changes with flake8 are pretty conservative imho.

> Depends on the view that you have. I see them more as TODOs for later since the "bad code" is in master at the moment anyways. But if you want to discuss that - sure.

I totally share this view, I think it's fine to use noqa as flag and come back later.

I admit you guys are 1000x more experts in this, but from what I can judge I think it's fine to merge this as first attempt and then in subsequent PRs improve and address flake 8 fails. 

Maybe worth to merge #1527 first? otherwise there will be potential conflicts more for phil.

flying-sheep commented:
Yes please! It’s been soooo long. I’ll happily fix the conflicts here as long as I don’t have to change the flit PR again.

Zethson commented:
@flying-sheep go ahead :)
I will patiently wait until your PR got merged.

ivirshup commented:
> I see them more as TODOs for later since the "bad code" is in master at the moment anyways. But if you want to discuss that - sure.

I think it's hard to tell the difference between a `noqa` that was added because: "most of the time, this rule is right, this time it is wrong" vs. "added to get rid of warning". Could any `noqa`s added in this PR get something searchable added to them (like `# noqa: {rule} TODO: fix me`) so we know why it was added?

In a future meeting we can discuss with the whole team how we will actually fix these. I think it's a good task for a hackathon/ sprint.

------------

Looking at this again, I think this is pretty close to done. Just a few documentation/ minor rule changes left.

- [x] Document how to turn off these checks in dev docs
- [x] Document the rules that are turned off (similar to [pandas](https://github.com/pandas-dev/pandas/blob/879cd22dd58b0574cdcaa7a26e396d5ec71a615a/setup.cfg#L71-L79))
- [x] Add autopep8 to precommit. If things can be fixed automatically, they should be. autopep8 should be able to get it's rules from  the flake8 config.
- [x] Add annotation to `noqa`s.
- [x] Also add `TODO` annotation to `except Exception`s that have been added.
- [x] Turn off E731
- [x] Turn off the rule that doesn't allow multiple leading `#` in comments
- [x] Turn off F811 for tests (rule violated by using fixture as test argument)
  - Possible solution: add noqa for this to all files under `tests`, separate check that all files under tests have this.
- [x] Ignore `build` `docs/_build` directories

Zethson commented:
> Could any noqas added in this PR get something searchable added to them (like # noqa: {rule} TODO: fix me) so we know why it was added?

The noqas already state what they are ignoring. I for now would not try to differentiate between noqas that we want to keep and noqas that we want to get rid of. We want to get rid of all of them in the follow up issue and only when examining all of them we will figure out which ones we want to keep.

> Document how to turn off these checks in dev docs

What do you mean? How to ignore a single line? How to fully ignore whole checks? I would always refer to the flake8 documentation, because it will certainly maintained better than the dev documentation.

> Add autopep8 to precommit. If things can be fixed automatically, they should be. autopep8 should be able to get it's rules from the flake8 config.

I wish it were that easy. autopep8 does not take its configuration from the flake8 config file nor can it fix all pep8 violations nor do Black and autopep8 always work nicely together. Black is an **opiniated** formatter. It formats consistently, but not necessarily compatible with other tools. I would not add autopep8, since I do not see any further benefit to the Black & flake8 combination, only more potential for issues and confused developers.

I agree with your other comments and will take care of them as soon as I got your answers :)

ivirshup commented:
> I for now would not try to differentiate between noqas that we want to keep and noqas that we want to get rid of. We want to get rid of all of them in the follow up issue and only when examining all of them we will figure out which ones we want to keep.

After this merges new ignore messages can be added for reasons like "this rule is generally good, but not in this specific case". Each of these will go through PR review, so will be vetted.

The ones added here largely have not been vetted, and are just being added so we don't get a failure. I would like to be able to distinguish between these cases. Once more `noqa` cases are added, it gets more complicated to find cases that haven't been vetted if they don't have some associated label.

--------------------------

> What do you mean? How to ignore a single line? How to fully ignore whole checks?

How to disable flake8 errors for a line or file.

> I would always refer to the flake8 documentation, because it will certainly maintained better than the dev documentation.

A link to the section of the flake8 docs on this would be great.

-------------------------

> I wish it were that easy. autopep8 does not take its configuration from the flake8 config file 

`autopep8` says it does this: https://github.com/hhatto/autopep8#configuration

> It formats consistently, but not necessarily compatible with other tools

I would like changes that are automatically applicable to be automatically applied. I'm thinking of things like white space in docstrings. Is there another way to automate these you can suggest?

---------

BTW, I've added a few more points to the checklist above. I would recommend trying to build the package and build the docs in the directory you're working in to see what files get generated so they can be added to the `ignore`.

flying-sheep committed:
Fix paga_path (#1047)

Fix paga_path

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

flying-sheep committed:
Restore setup.py

flying-sheep commented:
I reverted the setup.py change which hopefully means that an alrady-compliant setup.py from the flit PR will just pass [the check](https://github.com/theislab/scanpy/actions/runs/610117826) (once it runs, seems to be queued for a while now)

flying-sheep committed:
Go back to regular pip (#1702)

* Go back to regular flit

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson commented:
I'll fix this PR up in ~10 days. Appreciate your comments @ivirshup and think that some of your feedback could have been mitigated by me not going for short cuts :)

Will request (final hopefully) review then.

ivirshup committed:
codecov comment (#1704)

ivirshup committed:
Use joblib for parallelism in regress_out (#1695)

* Use joblib for parallism in regress_out

* release note

* fix link in release notes

* Add todo for resource test

pinin4fjords committed:
Add sparsificiation step before sparse-dependent Scrublet calls (#1707)

* Add sparsificiation step before sparse-dependent Scrublet calls

* Apply sparsification suggestion

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

* Fix imports

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

flying-sheep committed:
Fix version on Travis (#1713)

By default, Travis does `git clone --depth=50` which means the version can’t be detected from the git tag.

ivirshup committed:
`sc.metrics` module (add confusion matrix & Geary's C methods) (#915)

* Add `sc.metrics` with `gearys_c`

Add a module for computing useful metrics. Started off with Geary's C since I'm using it and finding it useful. I've also got a fairly fast way to calculate it worked out.

Unfortunatly my implementation runs into some issues with some global configs set by umap (see https://github.com/lmcinnes/umap/issues/306), so I'm going to see if that can be resolved before changing it.

* Add sc.metrics.confusion_matrix

* Better tests and output for confusion_matrix

* Workaround umap<0.4 and increase numerical stability of gearys_c

* Work around https://github.com/lmcinnes/umap/issues/306 by not
  calling out to kernel function. That code has been kept, but commented
  out.
* Increase numerical stability by casting data to system width. Tests
  were failing due to instability.

* Split up gearys_c tests

* Improved unexpected error message

* gearys_c working again. Sadly, a bit slower

* One option for doc strings

* Simplify implementation to use single dispatch

* release notes

ivirshup committed:
Fix clipped images in docs (#1717)

ivirshup committed:
Cleanup normalize_total (#1667)

* Cleanup normalize_total

* Add modification tests and copy kwarg for normalize_total

* Test that 'layers' argument is deprecated

* Added more mutation checks for normalize_total

* release note

* Error message

mjayasur committed:
deprecate scvi (#1703)

* deprecate scvi

* Update .azure-pipelines.yml

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

* remove :func: links to scvi in release notes

* remove tildes in front of scvi in release notes

* Update docs/release-notes/1.5.0.rst

Co-authored-by: Michael Jayasuriya <mjayasur@Michaels-MacBook-Air.local>
Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

alexmascension committed:
updated ecosystem.rst to add triku (#1722)

ivirshup committed:
Minor addition to contributing docs (#1726)

gokceneraslan committed:
Preserve category order when groupby is a list (#1735)

Preserve category order when groupby is a list

gokceneraslan committed:
Asymmetrical diverging colormaps and vcenter (#1551)

Add vcenter and norm arguments to plotting functions

Zethson committed:
rebased

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
rebasing

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
rebasing

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
rebasing

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
Merge branch 'master' into feature/flake8

Zethson committed:
add flake8 to dev docs

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add autopep8 to pre-commits

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add flake8 ignore docs

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add exception todos

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add ignore directories

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
reinstated lambdas

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix tests

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix tests

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix tests

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix tests

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix tests

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson commented:
@ivirshup I hope that I caught all of your comments.
The flake8 configuration is now minimal and pretty much only contains black violations or what you requested. I added and then removed autopep8 again, because it has other opinions on formatting than the opinionated formatter black. Yes, even with the flake8 configuration file. Black formatted the code then autopep8 and this cycle continues forever.
Added TODOs to exceptions and noqas are still easily searchable and mention what rule they ignore anyways.

I want to get this merged asap since the merge conflicts will just pile up.

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,20 @@
+         E126,  # continuation line over-indented for hanging indent -> black does not adhere to PEP8
+         E266,  # E266 too many leading '#' for block comment -> Scanpy allows them for comments into sections
+         E731   # Do not assign a lambda expression, use a def -> Scanpy allows lambda expression assignments
+exclude =
```
```suggestion
         E731,  # Do not assign a lambda expression, use a def -> Scanpy allows lambda expression assignments
         E741,  # allow I, O, l as variable names -> I is the identity matrix, i, j, k, l is reasonable indexing notation
exclude =
```

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -165,7 +165,6 @@ def dpt_timeseries(
 ):
     """\
     Heatmap of pseudotime series.
-
```
There should probably be a blank line here?

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,20 @@
+    __pycache__,
+    build,
+    docs/_build
+    dist
```
```suggestion
    docs/_build,
    dist,
per-file-ignores = 
    # F811 Redefinition of unused name from line, does not play nice with pytest fixtures
    tests/test*.py: F811,
```

Supposedly, you're not allowed to have comments on the same lines as entries (even though this clearly works above). See: https://flake8.pycqa.org/en/latest/user/configuration.html

ivirshup commented on scanpy/neighbors/__init__.py:
```diff
@@ -864,7 +860,7 @@ def _compute_connectivities_diffmap(self, density_normalize=True):
             if not self.knn:
                 mask = W > 1e-14
-                W[mask == False] = 0
+                W[mask == False] = 0  # noqa: E712
```
```suggestion
                W[~mask] = 0
```

ivirshup commented on scanpy/neighbors/__init__.py:
```diff
@@ -876,7 +872,7 @@ def _compute_connectivities_diffmap(self, density_normalize=True):
                             mask[j, i] = True
                 # set all entries that are not nearest neighbors to zero
-                W[mask == False] = 0
+                W[mask == False] = 0  # noqa: E712
```
```suggestion
                W[~mask] = 0
```

ivirshup commented on scanpy/neighbors/__init__.py:
```diff
@@ -1027,24 +1023,24 @@ def _get_dpt_row(self, i):
-                * (self.eigen_basis[i, l] - self.eigen_basis[:, l])
+                self.eigen_values[k]
+                / (1 - self.eigen_values[k])
+                * (self.eigen_basis[i, k] - self.eigen_basis[:, k])
```
This variable should probably be `j` if anything

ivirshup commented on .flake8:
```diff
@@ -0,0 +1,20 @@
+         E402,  # module level import not at top of file -> required to circumvent circular imports for Scanpys API
+         E126,  # continuation line over-indented for hanging indent -> black does not adhere to PEP8
+         E266,  # E266 too many leading '#' for block comment -> Scanpy allows them for comments into sections
+         E731   # Do not assign a lambda expression, use a def -> Scanpy allows lambda expression assignments
```
All of the comments need to be on separate lines from the ignore.

See:

* https://flake8.pycqa.org/en/latest/user/configuration.html
* https://docs.python.org/3/library/configparser.html#customizing-parser-behaviour
* https://docs.python.org/3/library/configparser.html#supported-ini-file-structure

It looks like this is what was causing you problems with autopep8, which should be fine to re-enable after this. It was parsing the config in a way that doesn't allow inline comments.

ivirshup commented on scanpy/get/get.py:
```diff
@@ -96,7 +96,7 @@ def rank_genes_groups_df(
     dim_df: pd.DataFrame,
     alt_index: pd.Index,
-    dim: "Literal['obs', 'var']",
+    dim: "Literal['obs', 'var']",  # noqa: F821
```
Could this be replaced with a `from ._compat import Literal` at the top of the file? Ditto for other occurrences.

Zethson replied:
Not working unfortunately.

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -62,7 +62,7 @@ def pca_overview(adata: AnnData, **params):
     if 'show' in params:
         del params['show']
-    scatterplots.pca(adata, **params, show=False)
+    scatterplots.pca(adata, **params, show=False)  # noqa: F821
```
Would importing `scatterplots` fix this?

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -967,7 +966,7 @@ def rank_genes_groups_violin(
         savefig_or_show(writekey, show=show, save=save)
         axs.append(_ax)
-    if show == False:
+    if not show:
```
This is a behavior change since `show` defaults to `False`

ivirshup commented on scanpy/plotting/_tools/paga.py:
```diff
@@ -148,7 +148,7 @@ def paga_compare(
         pl.suptitle(suptitle)
     _utils.savefig_or_show('paga_compare', show=show, save=save)
-    if show == False:
+    if not show:
```
Similarly, `show` defaults to `None` so this is a behavior change. I'm starting to think we should just ignore this rule.

Zethson replied:
While you are correct that this is a behavior change I am not really sure whether we should really compare to ==. It is considered "bad style". https://stackoverflow.com/questions/37103705/what-is-the-correct-way-to-check-for-false

"Falsy" value should be good enough here so I would just adhere to the standard.

Final choice is yours, but I would use "not".

ivirshup commented on scanpy/plotting/_utils.py:
```diff
@@ -708,9 +708,9 @@ def scatter_base(
+        left = panel_pos[2][2 * icolor]  # noqa: F841
         bottom = panel_pos[0][0]
-        width = draw_region_width / figure_width
+        width = draw_region_width / figure_width  # noqa: F841
```
Probably just delete definitions of `width` and `left`

ivirshup commented on scanpy/plotting/_utils.py:
```diff
@@ -940,8 +940,8 @@ def make_pos(pos, node=root, currentLevel=0, parent=None, vert_loc=0):
-        levels = {l: {TOTAL: levels[l], CURRENT: 0} for l in levels}
-    vert_gap = height / (max([l for l in levels]) + 1)
+        levels = {level: {TOTAL: levels[level], CURRENT: 0} for level in levels}
+    vert_gap = height / (max([level for level in levels]) + 1)
```
```suggestion
        levels = {k: {TOTAL: v, CURRENT: 0} for k, v in levels.items()}
    vert_gap = height / (max(levels.keys()) + 1)
```

ivirshup commented on scanpy/preprocessing/_deprecated/highly_variable_genes.py:
```diff
@@ -170,8 +170,10 @@ def filter_genes_dispersion(
+            df['dispersion'].values
+            - disp_mean_bin[  # use values here as index differs
+                df['mean_bin'].values
+            ].values
```
```suggestion
            # use values here as index differs
            df['dispersion'].values
            - disp_mean_bin[df['mean_bin'].values].values
```

Though the comment should probably go outside this expression.

I would also expect these are equivalent:

```python
disp_mean_bin[df['mean_bin'].values].values
disp_mean_bin.values[df['mean_bin']]
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -342,7 +342,7 @@ def check_adjacency(self):
             for iseg in range(self.segs_adjacency.shape[0]):
                 if n_edges_per_seg[iseg] == n_edges:
-                    neighbor_segs = self.segs_adjacency[iseg].todense().A1
+                    _ = self.segs_adjacency[iseg].todense().A1
```
Could these get `# noqa`s an `TODO`s? I'm assuming these should just be deleted, but am worried they're here for side effects.

ivirshup replied:
Similar for `prev_connecting_points`

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -903,9 +899,9 @@ def _detect_branching_single_haghverdi16(self, Dseg, tips):
         ps = [
             [0, 1, 2],  # start by computing distances from the first tip
-            [1, 2, 0],  #             -"-                       second tip
+            [1, 2, 0],  # -"-                       second tip
```
```suggestion
            [1, 2, 0],  #              -"-              second tip
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -903,9 +899,9 @@ def _detect_branching_single_haghverdi16(self, Dseg, tips):
+            [1, 2, 0],  # -"-                       second tip
             [2, 0, 1],
-        ]  #             -"-                       third tip
+        ]  # -"-                       third tip
```
Similar to above line, but moved up. Or revert these lines, if easier.

ivirshup commented on scanpy/tools/_rank_genes_groups.py:
```diff
@@ -37,6 +37,7 @@ def _ranks(X, mask=None, mask_rest=None):
     if issparse(X):
         merge = lambda tpl: vstack(tpl).toarray()
         adapt = lambda X: X.toarray()
+
```
??

Probably safe to revert the whole file.

ivirshup commented on scanpy/tools/_sim.py:
```diff
@@ -1077,16 +1074,6 @@ def sim_givenAdj(self, Adj: np.ndarray, model='line'):
-                sigma_noise=0.1,
-            )
-        ]
-
```
Ehh, I'd leave it. I'm not sure what happens in this file so I'd rather leave it untouched.

ivirshup commented on scanpy/tools/_paga.py:
```diff
@@ -534,24 +533,24 @@ def paga_compare_paths(
             ]
             path_mapped_orig_names = [
-                [orig_names1[int(s)] for s in l] for l in path_mapped
+                [orig_names1[int(s)] for s in link] for link in path_mapped
```
Could these get reverted? I'm not sure `link` is an appropriate name. I suspect these are `leafs`, but would rather have someone familiar with this code make that call.

Zethson replied:
They are redefinitions and should not stay as `l`. I'd be fine with either

1. renaming to _
2. renaming to k/j or something
3. Opening a follow up issue for this.

ivirshup replied:
I'm not seeing how these are redefinitions, as comprehensions get their own scope. Am I missing something? If anything, wouldn't `s` be the redefinition?

> renaming to _

This would be wrong, since the value is used, right?

> renaming to k/j or something

If there really is a problem with `l`, then I'm fairly confident this variable refers to `leaf`s. My preference is still not to touch it.

Zethson replied:
I honestly don't see it either, but Pycharm and flake8 agree that these are redefinitions. Maybe there's a global imported `l` somewhere. I went with leafs now.

ivirshup commented on scanpy/preprocessing/_simple.py:
```diff
@@ -752,7 +752,9 @@ def scale(
-    return scale_array(data, zero_center=zero_center, max_value=max_value, copy=copy)
+    return scale_array(
+        data, zero_center=zero_center, max_value=max_value, copy=copy  # noqa: F821
+    )
```
```suggestion
    if layer is not None:
        raise ValueError(f"`layer` argument inappropriate for value of type {type(X)}")
    if obsm is not None:
        raise ValueError(f"`obsm` argument inappropriate for value of type {type(X)}")
    return scale_array(X, zero_center=zero_center, max_value=max_value, copy=copy)
```

As a fallback, we'll assume`X` is array-like.

ivirshup commented on scanpy/external/pl.py:
```diff
@@ -332,15 +332,14 @@ def scrublet_score_distribution(
-    Plot histogram of doublet scores for observed transcriptomes and simulated doublets. 
+    Plot histogram of doublet scores for observed transcriptomes and simulated doublets.
+
+    The histogram for simulated doublets is useful for determining the correct doublet score threshold.
```
```suggestion
    The histogram for simulated doublets is useful for determining the correct doublet
    score threshold. 
```

Bit long, seemed fine before

ivirshup commented on docs/dev/code.rst:
```diff
@@ -16,7 +16,13 @@ Code style
+of this repository.
+To learn how to ignore checks per line please read
+`flake8 violations <https://flake8.pycqa.org/en/latest/user/violations.html>`__.
+Additionally, we use Scanpy’s
```
```suggestion
`Black <https://black.readthedocs.io/en/stable/the_black_code_style.html>`__
and
`flake8 <https://flake8.pycqa.org>`__.
We ignore a couple of flake8 checks which are documented in the .flake8 file in the root of this repository.
To learn how to ignore checks per line please read
`flake8 violations <https://flake8.pycqa.org/en/latest/user/violations.html>`__.
Additionally, we use Scanpy’s
```

ivirshup reviewed (changes requested):
Looks good! 

I left a lot of comments/ suggestions, but generally fall into a few categories.

* Places that I understand the code, I suggested style fixes (replacing `noqa`)
* Places where it looks like formatting is held over from mixing formatters/ rules we've abandoned. I've suggested reverting here.
* Fixes/ renamed variables in code I'm not familiar with, I've suggested reverting / adding a noqa.
* Renamed variables where I think the new name isn't appropriate. I've suggested reverting or a different name.
* There were a few places where replacing `var == False` with `not var` would be a behavior change since `var` defaults to `None`, `E712` is fine with `if var is False`.

-----------------------

Bigger points:

* I think the `autopep8` issues you were coming across were a config error. I've suggested a fix in the `.flake8` file. You should be able to turn on `autopep8` again after enabling that.
* Reminder on `# noqa`s, could all of the ones you've added get a `TODO` added inline? I want to be able to tell the difference between `noqa`s added now, and those added later with an editor search. Here, they may be hiding bugs. Later ones may be legit exceptions.

Zethson committed:
Add E741 to allowed flake8 violations.

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Add F811 flake8 ignore for tests

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Fix mask comparison

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Fix mask comparison

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
fix flake8 config file

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
readded autopep8

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
import Literal

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
revert literal import

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix scatterplot pca import

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
false comparison & unused vars

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
Add cleaner level determination

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Fix comment formatting

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Add smoother dev documentation

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
fix flake8

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
Readd long comment

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
Assuming X as array like

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
fix flake8

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix flake8 config

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
reverted rank_genes

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
fix disp_mean_bin formatting

Co-authored-by: Isaac Virshup <ivirshup@gmail.com>

Zethson committed:
fix formatting

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
add final todos

Signed-off-by: Zethson <lukas.heumos@posteo.net>

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -365,7 +365,7 @@ def _fig_show_save_or_axes(plot_obj, return_fig, show, save):
         savefig_or_show(plot_obj.DEFAULT_SAVE_PREFIX, show=show, save=save)
         show = settings.autoshow if show is None else show
-        if not show:
+        if show == False:  # noqa: E712  # TODO Does this need fixing? -> is False
```
I'd be fine with this:

```suggestion
        if show is False:
```

ivirshup replied:
I think `flake8` allows this, but pep8 doesn't. Not sure if this would create conflicts with `autopep8`.

ivirshup commented on scanpy/plotting/_tools/__init__.py:
```diff
@@ -967,7 +967,7 @@ def rank_genes_groups_violin(
         savefig_or_show(writekey, show=show, save=save)
         axs.append(_ax)
-    if show == False:
+    if show == False:  # noqa: E712  TODO Does this need fixing?
```
```suggestion
    if show is False:
```

ivirshup commented on scanpy/plotting/_tools/paga.py:
```diff
@@ -148,7 +148,7 @@ def paga_compare(
         pl.suptitle(suptitle)
     _utils.savefig_or_show('paga_compare', show=show, save=save)
-    if not show:
+    if show == False:  # noqa: E712  # TODO Does this need fixing?
```
```suggestion
    if show is False:
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -899,9 +903,9 @@ def _detect_branching_single_haghverdi16(self, Dseg, tips):
             [0, 1, 2],  # start by computing distances from the first tip
-            [1, 2, 0],  # -"-                       second tip
+            [1, 2, 0],  #             -"-                       second tip
             [2, 0, 1],
```
```suggestion
            [2, 0, 1],  #             -"-                       third tip
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -899,9 +903,9 @@ def _detect_branching_single_haghverdi16(self, Dseg, tips):
+            [1, 2, 0],  #             -"-                       second tip
             [2, 0, 1],
-        ]  # -"-                       third tip
+        ]  #             -"-                       third tip
```
```suggestion
        ]
```

ivirshup commented on scanpy/get/get.py:
```diff
@@ -6,7 +6,6 @@
 from scipy.sparse import spmatrix
 
 from anndata import AnnData
```
```suggestion
from anndata import AnnData

from .._compat import Literal
```

ivirshup commented on scanpy/get/get.py:
```diff
@@ -96,7 +95,7 @@ def rank_genes_groups_df(
     dim_df: pd.DataFrame,
     alt_index: pd.Index,
-    dim: "Literal['obs', 'var']",
+    dim: "Literal['obs', 'var']",  # noqa: F821  # TODO Does this need fixing?
```
```suggestion
    dim: Literal['obs', 'var'],
```

ivirshup commented on scanpy/get/get.py:
```diff
@@ -176,7 +175,7 @@ def _get_array_values(
     dim_names: pd.Index,
     keys: List[str],
-    axis: "Literal[0, 1]",
+    axis: "Literal[0, 1]",  # noqa: F821  # TODO Does this need fixing?
```
```suggestion
    axis: Literal[0, 1],
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -984,7 +984,7 @@ def __detect_branching_haghverdi16(
             )
         # init list to store new segments
-        ssegs = []
+        ssegs = []  # noqa: F841
```
```suggestion
        ssegs = []  # noqa: F841 # TODO: Look into this
```

ivirshup commented on scanpy/tools/_dpt.py:
```diff
@@ -743,11 +747,7 @@ def _detect_branching(
-        List[List[int]],
-        List[List[int]],
-        int,
+        List[np.ndarray], List[np.ndarray], List[List[int]], List[List[int]], int
```
```suggestion
        List[np.ndarray],
        List[np.ndarray],
        List[List[int]],
        List[List[int]],
        int,
```

I think this is fine. Seems intentional, due to the trailing comma.

ivirshup reviewed (changes requested):
Thanks for the changes, just a few minor things I've added suggestions for and the `paga` variables.

Zethson committed:
boolean checks with is

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
_dpt formatting

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
literal fixes

Signed-off-by: Zethson <lukas.heumos@posteo.net>

Zethson committed:
links to leafs

Signed-off-by: Zethson <lukas.heumos@posteo.net>

ivirshup committed:
revert paga variable naming

ivirshup reviewed (approved):
Looks good, thanks for all the work!

We should add a release note for this at some point, I'm just not sure where yet, probably a section for dev practices. Could you suggest a line for that?

I was unsure about the variable naming for PAGA, so I've decided to revert that. I couldn't get flake8 to call it a redefinition.

Zethson commented:
> Looks good, thanks for all the work!
> 
> We should add a release note for this at some point, I'm just not sure where yet, probably a section for dev practices. Could you suggest a line for that?
> 
> I was unsure about the variable naming for PAGA, so I've decided to revert that. I couldn't get flake8 to call it a redefinition.

:tada: 
Maybe "Enabled flake8 (https://flake8.pycqa.org/en/latest/) pre-commit to run code style checks"?
Everything else might just be details that people will uncover anyways since the workflows might complain :)
